### DISCUSSION TITLE
Clearer handling of application initialization

### DISCUSSION THREAD

**Discussion by atombender** [vercel/next.js#8969] [Help]
# Feature request

## Is your feature request related to a problem? Please describe.

There are numerous examples of how to "bootstrap" an app with state: Redux, Apollo, authentication, and so on.

These typically rely on "augmenting" the app with various HOC wrappers. For example, `next-redux-wrapper` provides a wrapper that injects itself into `App`, whereas `examples/with-apollo-auth` provides a wrapper that must be wrapped around every single page.

They all have various solutions for the same thing: Initializing global state, storing it in the context, handling hydration/dehyration, plus using heuristics like `typeof window !== 'undefined'` to conditionally store the state in locally or in a global variable (e.g. for the Apollo client or Redux store) for when it is used client-side.

It's messy. 

My use case: I am using Apollo, and I want Apollo to be available both as context (so I can use components like `Query`, and `useApolloClient()`), and I need it in every page's `getInitialProps` because that's where the translation of query parameters to actual data should be done:

```javascript
ArticlePage.getInitialProps = async ({ctx, query}): Promise<IArticlePage> => {
  return {
    user: ctx.apolloClient.query({query: gql`{ user { id, name } }`}),
    article: ctx.apolloClient.query({
      query: gql`query Article($id: String!) {
        article { id, name }
      }`,
      variables: {id: query.id}
    }),
  }
}
```

I _also_ need to load some global data such as the user's account, which belongs in `App`. I also have some global data such as the current URL, extracted from `ctx.req`, that I need to share as context. (The current URL also needs to be used to initialize Apollo, coincidentally.)

The question is _where_ to initialize it.

I don't like explicitly wrapping every page component (as in `examples/with-apollo-auth`) for various reasons: It's a violation of DRY; it adds an unnecessary implementation detail to every page; it adds initialization logic at the page level instead of at the application level; and it's something that needs to be done identically in every single page file (makes it easy to forget).

Initially, I followed the `next-redux-wrapper` method and tried replacing `App` itself with one that loads the Apollo client (`export withApollo(App)`). However, the page's `getInitialProps` runs _before_ `App`, so it does not get access to the Apollo client!

So I tried the `with-apollo-auth` method and wrapped each page instead (`export withApollo(MyPage)`). But this has the issue that Apollo only becomes available inside the page. `App` does not ever get access to the Apollo context. That means `App` can no longer be used as a place to insert global components that require Apollo. They all have to be pushed down to the page level.

From my perspective, Next.js has a big gap in functionality here. There's no single point where this sort of initialization can be handled for _both_ `App` and individual pages. There's also this implicit contract that each page needs to handle the transition from server-side to client-side, when Next.js itself should be the one driving this.

In other words, "app-level" concerns (as opposed to page/component-level) have no explicit flow. `App` seems a bit misnamed; it's initialized after the page, but at the same time rendered around the page; it doesn't seem to represent the _app_ as much as the outer container. (This is exemplified by the fact that `App` in fact has `getInitialProps` called on every router navigation, meaning that that method has to be lightweight and must be careful to not, say, load the user's account from the database on every call.) It's where one should _apparently_ place global concerns, but not all global concerns can be placed there.

I don't see what a clean solution might be for my use case, with today's Next.js, other than wrapping _both_ individual pages and `App`. Maybe I'm missing something. But this stuff shouldn't be so hard. I shouldn't have to write complicated HOCs just to initialize some data on startup. It shouldn't be necessary to wrap pages.

## Describe the solution you'd like

There should be a single place to initialize the application as a whole. It should execute before the app is mounted as a React app, allowing one to write logic that can inspect `req` and the application context in a single location before a page is mounted, both server- and client-side.

If initialization was made more explicit, it would also be a way to allow a "middleware" system. Look at all the permutations of `with-apollo-*` or `with-redux-*` examples. Look at `with-apollo-and-redux-saga/lib/withApollo.js`. Why isn't this just a middleware that inserts itself into the initialization stack and gives you an Apollo client and a Redux store?

**Comment by timneutkens** [comment]
> Initially, I followed the next-redux-wrapper method and tried replacing App itself with one that loads the Apollo client (export withApollo(App)). However, the page's getInitialProps runs before App, so it does not get access to the Apollo client!

This is incorrect, Next.js only calls `_app`'s getInitialProps which implements calling getInitialProps on the page through `await Component.getInitialProps` (metacode).

**Comment by tettoffensive** [comment]
@atombender Ever find a way to do this?

**Comment by aashish254** [comment]
**# Solution: Application Initialization Patterns in Next.js**\n\nGreat question! You're absolutely right that the current patterns are fragmented and force you to choose between wrapping every page vs managing everything at App level. Let me share some proven solutions based on your Apollo use case.\n\n## Current State (Your Problem)\n\nYou've identified the core issue perfectly: **there's no single initialization point** that works for both `App` AND individual pages' `getInitialProps`. The current contract is indeed confusing.\n\n## Solution 1: Use App + getInitialProps Strategy (Recommended)\n\nThe key insight is that `_app.getInitialProps` runs BEFORE the page component, so you CAN initialize there:\n\n```javascript\n// pages/_app.js\nimport { ApolloProvider } from '@apollo/client';\nimport { createApolloClient } from '../lib/apollo';\n\nfunction MyApp({ Component, pageProps, ctx }) {\n  // Initialize global Apollo client here\n  const apolloClient = createApolloClient(ctx.req);\n  \n  return (\n    <ApolloProvider client={apolloClient}>\n      <Component {...pageProps} />\n    </ApolloProvider>\n  );\n}\n\nMyApp.getInitialProps = async (ctx) => {\n  // THIS RUNS FIRST and can access ctx.req!\n  const appProps = await App.getInitialProps(ctx);\n  \n  // NOW you have access to BOTH req AND context\n  // Initialize global data here (user account, current URL, etc.)\n  const userData = ctx.req ? getUserFromDb(ctx.req) : null;\n  const currentUrl = getUrlFromReq(ctx.req);\n  \n  // Pass down to pages via pageProps\n  return {\n    ...appProps,\n    initialUserData: userData,\n    initialUrl: currentUrl,\n  };\n};\n\nexport default MyApp;\n```\n\nThen in your pages:\n```javascript\n// pages/article/[id].js\nfunction ArticlePage({ initialUserData, initialUrl }) {\n  // Both available here during SSR and client-side!\n  const user = useQuery(GET_USER, { variables: { ... } });\n  \n  return <div>{/* render */}</div>;\n}\n\nArticlePage.getInitialProps = async ({ query }) => {\n  return { articleId: query.id };\n};\n```\n\n## Solution 2: Custom Context Provider Pattern\n\nFor more complex initialization flows, create a custom context:\n\n```javascript\n// contexts/AppContext.js\nimport { createContext, useContext } from 'react';\n\nexport const AppContext = createContext(null);\n\nexport function useAppContext() {\n  const context = useContext(AppContext);\n  if (!context) throw new Error('useAppContext must be used within AppProvider');\n  return context;\n}\n\n// pages/_app.js\nexport default function MyApp({ Component, pageProps, ctx }) {\n  const [appData, setAppData] = useState(null);\n  \n  useEffect(() => {\n    // Run initialization here during client-side mount\n    initAppData();\n  }, []);\n  \n  return (\n    <AppContext.Provider value={{ ...appData, apolloClient }}>\n      <Component {...pageProps} />\n    </AppContext.Provider>\n  );\n}\n```\n\n## Solution 3: Middleware Pattern (Next.js 12+)\n\nIf you're on Next.js 12+, you should use **middleware** instead:\n\n```javascript\n// middleware.js (automatically runs before every request!)\nimport { NextResponse } from 'next/server';\n\nexport function middleware(request) {\n  // Initialize global data\n  const userData = getUserFromRequest(request);\n  const currentUrl = request.url;\n  \n  // Set as headers or cookies\n  const response = NextResponse.next();\n  response.headers.set('x-user-id', userData.id);\n  \n  return response;\n}\n```\n\nThis solves your problem elegantly without wrapper complexity!\n\n## Best Practice Recommendation\n\nFor your Apollo + user data use case, I recommend:\n\n1. **Use `_app.getInitialProps`** for initializing the Apollo client and shared global data\n2. **Pass initialization data via `pageProps`** to make it available in all pages\n3. **Use React Query or SWR** instead of Apollo for simpler hydration (both handle server/client differently)\n4. **Consider Next.js 13 App Router** which has built-in data fetching that eliminates most of these problems\n\n## Why This Works\n\n- ✅ Single initialization point in `_app.getInitialProps`\n- ✅ Can access `ctx.req` server-side\n- ✅ Data propagates to ALL pages via pageProps\n- ✅ No need to wrap individual pages\n- ✅ Follows DRY principle\n- ✅ Clear separation: App handles global concerns, pages handle local\n\nThe pattern is clean once you understand the execution order:\n1. Server: `req -> _app.getInitialProps -> page.getInitialProps -> render`\n2. Client: `hydration -> _app mount -> page mount`\n\nHope this helps solve your initialization challenge! 🚀
